### DISCUSSION TITLE
Using Tx & Rx pins in the SH-Esp32 with the engine hat

### DISCUSSION THREAD

**Discussion by iuriaranda** [hatlabs/discussions#44] [Q&A] [answered]
Hi,

I want to connect an ultrasonic tank level sensor to the SH-ESP32, which has the engine hat installed. The sensor is the DS1603L.V1 and uses two serial wires (Tx & Rx) for communication. The problem is that I'm not sure to which pins I need to connect those wires. From the docs I see that GPIO 1&3 are serial Tx & Rx pins, but I don't understand if they are already used or if they are available and can be used for this purpose. Specially if the engine hat is mounted, as it obstructs those pins in the GPIO header.

How do I best approach this?

Thanks!

**Comment by mairas** [comment, marked as answer]
The ESP32 has a nice IO MUX feature, letting you connect most digital peripherals to any GPIO pin. There are ~two~ three UART (serial) peripherals. The first one is by default connected to GPIOs 1 and 3 and on the dev boards goes to USB to serial IC. Better not to touch that, or you lose easy programming and serial debugging facilities. The two other UARTs, however, are free to grab and use with whatever pins you like. If you need more, the SoftwareSerial library provides a bit banging serial interface that works fine at lower speeds.

Here's one example for initializing `Serial1` on custom pins (lines 27 & 28):

https://github.com/SensESP/NMEA0183/blob/main/examples/gnss_receiver.cpp

**Reply by iuriaranda** [reply]
Oh interesting! I didn't know that. 

Thanks, that solves my problem :slightly_smiling_face:

**Reply by iuriaranda** [reply]
One more question. Apparently the accuracy of the sensor varies depending on the supplied voltage, and in order to get readings through a stainless steel tank I'm trying to run it from 12V. Would I still be able to read the serial output of the sensor from the SH-ESP32 if it's outputting 12V?

**Reply by mairas** [reply]
Hi, didn't notice this follow-up question earlier... Just like @hwater mentioned, you need to make sure the serial signal output remains under 3.3V. If you're not using all digital inputs on the engine hat, though, you could actually use one of its digital inputs for the serial RX! At least for slower bit rates.

**Comment by hwater** [comment]
Be carefull with any higher voltages than 3.3 Volt on ESP32 or Raspberry Pins. They don't like this at all.
